Exception-Handling.md

```diff
@@ -0,0 +1,209 @@
+
+Ruby, like many programming languages chooses to handle errors with a technique called *Exception Handling*.  When an error occurs, Ruby generates an object instance called an Exception.  Exceptions are always subclasses of the [Exception](https://ruby-doc.org/core-2.3.1/Exception.html) class. 
+
+[![exceptions](images/exceptions.png)](http://findnerd.com/list/view/Exception-Handling-in-Rails-using-begin-rescue/21677/)
+
+You can mark a section to recover from an error with a rescue block.  If an Exception is not "rescued" it will end the program with an error message.  
+
+Here is an example of an error being handled:
+
+```ruby
+# sample_exception.rb
+begin
+  # Dividing by zero causes an error
+  quotient = 5 / 0
+  puts "Made it past the error"
+rescue
+  quotient = nil
+  puts "Rescued the error and set quotient to nil"
+end
+```
+
+The output would be:
+```bash
+$  ruby sample_exception.rb
+Rescued the error and set quotient to nil
+```
+
+Without the rescue:
+
+```ruby
+# sample_exception.rb
+begin
+  # Dividing by zero causes an error
+  quotient = 5 / 0
+  puts "Made it past the error"
+end
+```
+The output would be:
+```bash
+$  ruby sample_exception.rb
+basic_exception_without_rescue.rb:4:in `/': divided by 0 (ZeroDivisionError)
+	from basic_exception_without_rescue.rb:4:in `<main>'
+```
+
+### Rescuing Specific Types of Exceptions
+
+There are a number of kinds of exceptions both for [regular'ol Ruby](http://blog.nicksieger.com/articles/2006/09/06/rubys-exception-hierarchy/) & [Rails](https://ruby-doc.org/core-2.3.1/Exception.html).  Some types of errors we probably don't want to try to rescue because they are irrecoverable or so unexpected that we shouldn't write code to recover from them.  We can however set up specific blocks to recover from specific types of errors that we can and want to recover from.  Generally recoverable errors inherit from `StandardError` and for this reason when you place a rescue block without specifying the exception class Ruby will only handle exceptions inheriting from StandardError.  
+
+The syntax will look like this:
+
+```ruby
+begin
+  # some code that will create an exception.
+  
+rescue <some kind of exception>
+  # Code to recover from that exception
+
+rescue <some other kind of exception>
+  # Code to recover from another type of exception
+
+rescue
+  # other exceptions handled here
+
+ensure
+  #  This code runs no matter what.  
+
+end 
+```
+#### One Ruby Example - CSVReader
+
+In Ruby we could write code to handle problems, like in our CSV code to deal with problems, like being unable to open the CSV file.
+
+```ruby
+content = []
+begin
+  CSV.foreach("file.csv") do |row|
+
+    content << row
+  end
+
+  rescue NameError
```

Comment:
`NameError` is only raised in cases where our code refers to a local variable or method that is not present in the current scope.

In this case we'll be getting a different sort of exception, coming from an OS-level error. We can handle it with this code:

```ruby
begin
  CSV.foreach("file.csv") do |row|
    content << row
  end
rescue SystemCallError
  puts "I couldn't open the file."
end
```